### ISSUE TITLE
Request: Leaner tree view

### ISSUE THREAD

**Issue by toxxmeister** [ForkIssues/Tracker#240]
I prefer to use tree view for changes, however it can become quite heavy on horizontal space in certain projects, such as Java ones, where you have large amounts of nesting due to how directories are organized.
It'd be great if the UI could squash some of these directory labels when it makes sense.

Here is a visual example of what I have in mind:

Current tree view:
```
+ app
    + src
        + main
            + java
                + com
                    + mycompany
                        + myproduct
                            + package
                                * AFile.java
                                + subpackage
                                    * AnotherFile.java
```

Leaner tree view:
```
+ app/src/main/java/com/mycompany/myproduct/package
    * AFile.java
    + subpackage
        * AnotherFile.java
```

**Comment by DanPristupov** [comment]
1. This isn't a trivial problem.
Guess you have an array of changed file paths (same as `git status`).
First you need to transform it to a nested structure.
Then you need to calculate a difference between the existing treeview structure and find out which items were removed and which were added. Paul Heckel's diff algorithm should probably be used there.
Some new files can appear in the middle of a 'squashed' folder. You need to calculate which folders must be 'unsquashed' and rework the tree structure to apply the diff.
Then you need to squash folders which are not used anymore.

The algorithm should be able to handle at least 5000 items without a delay, so complexity O(n) must be logarithmic or linear.

It's possible, but I'm not sure if it's worth spending time.

2. Do you know any IDE which has such a tree view? Spending a lot of time on a feature I never saw in RL seems to be risky.

**Comment by DanPristupov** [comment]
As an alternative, I can decrease the offset for nested items and make treeview more compact. Like on the screenshot:
![2018-03-07 at 18 40](https://user-images.githubusercontent.com/618115/37108330-1d972d3a-2237-11e8-87cc-e5c1c044ac35.png)

**Comment by nebhale** [comment]
I'm _not_ a tree user in Fork, but this is definitely a thing in other tree views for Java projects.  If you're not aware, Java projects are namespaced with "packages".  Packages are a folder structure, and typically follow reverse domain names.  So it's really common to see something `org.cloudfoundry.reactor.client.v3.applications` and the first three folders be completely empty.  Java IDE's especially, typically compress these into a single folder level (some even elide the names a bit like `o.c.r.client.v3.applications`).  Something like the following:

![screen shot 2018-03-07 at 11 35 04](https://user-images.githubusercontent.com/60754/37113903-a14a0382-21fb-11e8-94ad-d4faa9f36fc3.png)

In that example, you can see that compression can happen at any level (any time there is an empty directory, it can be compressed upwards) and that compression can happen next to uncompressed directories and files.  The choice to use dots is to meet Java package name expectations, but an equivalent feature could easily write `org/cloudfoundry/reactor` instead and then begin the tree there.

As far as an algorithm, it appears (based on a [popular issue](https://github.com/atom/tree-view/issues/125#issuecomment-284130284)) that Atom has an option to "Collapse Directories".  I'm guessing that they've got similar performance requires as Fork does (5000 items without delay), so their code might show you the algorithm they use to implement it.

**Comment by toxxmeister** [comment]
The screenshot provided by @nebhale showcases exactly what I meant.
Tree view can become quite unusable in large Java projects, where making changes in packages far apart from each other results in a lot of wasted vertical and horizontal space.

**Comment by DanPristupov** [comment]
@nebhale no, I didn't know that java projects have such a structure. It all makes sense now, thank you for the clarification.

Regarding the algorithm. I briefly checked the link you mentioned and noticed that this feature is not enabled by default in Atom, because it has exactly the problem which I mentioned:
> Some new files can appear in the middle of a 'squashed' folder. You need to calculate which folders must be 'unsquashed' and rework the tree structure to apply the diff.
Then you need to squash folders which are not used anymore.

https://github.com/atom/tree-view/issues/1015 

Anyway, an option to collapse empty folders should improve life for Java programmers a lot. I'll figure out how to implement that.

**Comment by DanPristupov** [comment]
IDE always have many changed files, but it can be a bit more complicated with git client.

What behavior would you expect if there's just a **single** modified file:
`org/cloudfoundry/reactor/client/v3/applications/path/to/file.txt`?
Showing a huge `org/cloudfoundry/reactor/client/v3/applications/path/to` folder seems to be really confusing.

**Comment by nebhale** [comment]
Maybe personal preference, but I prefer the flattened structure (note that I was not always in this camp, only the last decade or so).

![screen shot 2018-03-07 at 13 52 05](https://user-images.githubusercontent.com/60754/37120293-d4b9d252-220e-11e8-8d65-ff8b1e337b7b.png)

**Comment by justinsyc** [comment]
Will really love this feature. It is crucial for Java development.

**Comment by toxxmeister** [comment]
@DanPristupov 
Any update on whether this will be considered?
It's currently my biggest pain point with an otherwise superb app.
